pyanaconda/modules/timezone/geoloc.py

```diff
@@ -0,0 +1,132 @@
+# Public License for more details.  You should have received a copy of the
+# GNU General Public License along with this program; if not, write to the
+# Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301, USA.  Any Red Hat trademarks that are incorporated in the
+# source code or documentation are not subject to the GNU General Public
+# License and may only be used or replicated with the express permission of
+# Red Hat, Inc.
+#
+from abc import abstractmethod
+import requests
+from pyanaconda.core import constants
+from pyanaconda.timezone import get_preferred_timezone, is_valid_timezone
+from pyanaconda.modules.common.structures.timezone import GeolocationData
+
+from pyanaconda.anaconda_loggers import get_module_logger
+log = get_module_logger(__name__)
+
+
+class GeolocationBackend:
+    """Base class for GeoIP backends."""
+    @property
+    @abstractmethod
+    def name(self):
+        """Get name of the backend
+
+        :return str: name of the backend
+        """
+        return ""
+
+    @abstractmethod
+    def locate(self):
+        """Download the geolocation data.
+
+        Subclasses implement this method with actual backend-specific code.
+
+        :return GeolocationData: data structure describing geolocation results
+        """
+        pass
+
+    def __str__(self):
+        return self.name
+
+
+class FedoraGeoIPProvider(GeolocationBackend):
+    """The Fedora GeoIP service provider."""
+    @property
+    def name(self):
+        return "Fedora GeoIP"
+
+    def locate(self):
+        try:
+            reply = requests.get(
+                "https://geoip.fedoraproject.org/city",
+                timeout=constants.GEOLOC_CONNECTION_TIMEOUT,
+                verify=True
+            )
+            if reply.status_code == requests.codes.ok:  # pylint: disable=no-member
+                json_reply = reply.json()
+                territory = json_reply.get("country_code", "")
+                timezone = json_reply.get("time_zone", "")
+
+                # check if the timezone returned by the API is valid
+                if not is_valid_timezone(timezone):
+                    # try to get a timezone from the territory code
+                    timezone = get_preferred_timezone(territory)
+
+                if territory or timezone:
+                    return GeolocationData.from_values(
+                        territory=territory,
+                        timezone=timezone,
+                    )
+            else:
+                log.error("Geoloc: Fedora GeoIP API lookup failed with status code: %s",
+                          reply.status_code)
+
+        except requests.exceptions.RequestException as exc:
+            log.debug("Geoloc: RequestException for Fedora GeoIP API lookup:\n%s", exc)
+        except ValueError as exc:
+            log.debug("Geoloc: Unable to decode GeoIP JSON:\n%s", exc)
+
+        return None
```

Comment:
This is up to you, but I would always return some geolocation data even if there is nothing set. In general, it is a good practice that prevents issues with None.

Reply:
Sounds good, one more argument for a helper property to decide if GeolocationData is valid.